asserts/tool/sign.go

```diff
@@ -0,0 +1,249 @@
+	// "body": used as the body of the assertion
+	Statement []byte
+
+	// The revision of the new assertion
+	Revision int
+}
+
+func parseStatement(req *SignRequest, dest interface{}) error {
+	switch req.StatementMediaType {
+	case YAMLInput:
+		err := yaml.Unmarshal(req.Statement, dest)
+		if err != nil {
+			return fmt.Errorf("cannot parse the assertion input as YAML: %v", err)
+		}
+	case JSONInput:
+		dec := json.NewDecoder(bytes.NewBuffer(req.Statement))
+		// we want control over supporting only integers
+		dec.UseNumber()
+		err := dec.Decode(dest)
+		if err != nil {
+			return fmt.Errorf("cannot parse the assertion input as JSON: %v", err)
+		}
+	default:
+		return fmt.Errorf("unsupported media type for assertion input: %q", req.StatementMediaType)
+	}
+	return nil
+}
+
+type nestedStatement struct {
+	Headers map[string]interface{} `yaml:"headers" json:"headers"`
+	Body    string                 `yaml:"body" json:"body"`
+}
+
+// Sign produces the text of a signed assertion as specified by req.
+func Sign(req *SignRequest, keypairMgr asserts.KeypairManager) ([]byte, error) {
+	typ := asserts.Type(req.AssertionType)
+	if typ == nil {
+		return nil, fmt.Errorf("invalid assertion type: %q", req.AssertionType)
+	}
+	if req.Revision < 0 {
+		return nil, fmt.Errorf("assertion revision cannot be negative")
+	}
+	if req.AccountKey == nil && req.KeyID == "" {
+		return nil, fmt.Errorf("both account-key and key id were not specified")
+	}
+
+	var nestedStatement nestedStatement
+	err := parseStatement(req, &nestedStatement)
+	if err != nil {
+		return nil, err
+	}
+	if nestedStatement.Headers == nil {
+		// flat headers, reparse
+		err := parseStatement(req, &nestedStatement.Headers)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	headers, err := stringify(nestedStatement.Headers)
+	if err != nil {
+		return nil, err
+	}
+	body := []byte(nestedStatement.Body)
+
+	keyID := req.KeyID
+	authorityID := req.AuthorityID
+
+	if req.AccountKey != nil {
+		if keyID != "" || authorityID != "" {
+			return nil, fmt.Errorf("cannot mix specifying an account-key together with key id and/or authority-id")
+		}
+
+		// use the account-key as a handle to get the information about
+		// signer and key id
+		a, err := asserts.Decode(req.AccountKey)
+		if err != nil {
+			return nil, fmt.Errorf("cannot parse handle account-key: %v", err)
+		}
+		accKey, ok := a.(*asserts.AccountKey)
+		if !ok {
+			return nil, fmt.Errorf("cannot use handle account-key, not actually an account-key, got: %s", a.Type().Name)
+		}
+		keyID = accKey.PublicKeyID()
+		authorityID = accKey.AccountID()
+
+		// extra sanity checks about fingerprint
+		pk, err := keypairMgr.Get(authorityID, keyID)
+		if err != nil {
+			return nil, err
+		}
+		expFpr := accKey.PublicKeyFingerprint()
+		gotFpr := pk.PublicKey().Fingerprint()
+		if gotFpr != expFpr {
+			return nil, fmt.Errorf("cannot use found private key, fingerprint does not match account-key, expected %q, got: %s", expFpr, gotFpr)
+		}
```

Comment:
Checks above also ought to be done by Database.Sign, and per our conversation today we shouldn't ask the user for the account key, but should rather have it at hand in our database (fetching it if necessary, at some point).